File: LLD_Java/BookMyShow/BookMyShow.java
package Java.BookMyShow;

import Java.BookMyShow.Movie.Movie;
import Java.BookMyShow.Movie.MovieController;
import Java.BookMyShow.Theatre.TheatreController;
import Java.BookMyShow.Model.*;
import Java.BookMyShow.Theatre.*;
import Java.BookMyShow.Booking.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class BookMyShow {
    MovieController movieController;
    TheatreController theaterController;
    BookMyShow(){
        movieController = new MovieController();
        theaterController = new TheatreController();       
    }

    private void createBooking(Location location, String movieName){

        //1. search movie by my location
        List<Movie> movieList = movieController.getMoviesByLocation(location);

        //2. select the movie which you want to see. i want to see Baahubali
        Movie targetMovie = null;
        for (Movie movie: movieList){
            if (movie.getMovieName().equals(movieName)){
                targetMovie = movie;
            }
        }

        //3. get all show of this movie in given location
        Map<Theatre, List<Show>> showsTheatreWise = theaterController.getAllShow(targetMovie, location);

        //4. select the particular show user is interested in
        List<Show> runningShows = null;
        Show interestedShow = null;
        for(Map.Entry<Theatre,List<Show>> entry : showsTheatreWise.entrySet()) {
            runningShows = entry.getValue();
            interestedShow = runningShows.get(0);
            break;
        }

        // Map.Entry<Theatre,List<Show>> entry = showsTheatreWise.entrySet().iterator().next();
        // List<Show> runningShows = entry.getValue();
        // Show interestedShow = runningShows.get(0);

        //5. select the seat
        int seatNumber = 30;
        List<Integer> bookedSeats = interestedShow.getBookedSeatIds();
        if(!bookedSeats.contains(seatNumber)){
            bookedSeats.add(seatNumber);
            //startPayment
            Booking booking = new Booking();
            List<Seat> myBookedSeats = new ArrayList<>();
            for(Seat screenSeat : interestedShow.getScreen().getSeats()) {
                if(screenSeat.getSeatID() == seatNumber) {
                    myBookedSeats.add(screenSeat);
                }
            }
            booking.setBookedSeats(myBookedSeats);
            booking.setShow(interestedShow);
        } else {
            //throw exception
            System.out.println("seat already booked, try again");
            return;
        }

        System.out.println("BOOKING SUCCESSFUL");
    }

    private void createMovies() {

        Location lucknow = createLocation(123, "Jankipuram", "Lucknow", "UP", "INDIA");
        Location bangalore = createLocation(453, "Whitefield", "Bangalore", "Karnatak", "INDIA");
        //create Movies1
        Movie movie1 = new Movie();
        movie1.setMovieID(1);
        movie1.setMovieName("movie1");
        movie1.setMovieDurationInMin(128);

        //create Movies2
        Movie movie2 = new Movie();
        movie2.setMovieID(2);
        movie2.setMovieName("movie2");
        movie2.setMovieDurationInMin(180);


        //add movies against the cities
        movieController.addMovie(movie1, bangalore);
        movieController.addMovie(movie1, lucknow);
        movieController.addMovie(movie2, bangalore);
        movieController.addMovie(movie2, lucknow);
    }


    private Location createLocation(int pincode, String address, String city, String state, String country){
        Location location = new Location(pincode, address, city, state, country);
        return location;
    }

    private List<Seat> createSeats(){
        List<Seat> seats = new ArrayList<>();

        // 1 to 30
        for (int i =0; i<=30; i++){
            Seat seat = new Seat();
            seat.setSeatType(SeatType.BASIC);
            seat.setSeatID(i);
            seats.add(seat);
        }
        // 31 to 60
        for (int i = 31; i<=60; i++){
            Seat seat = new Seat();
            seat.setSeatType(SeatType.PREMIUM);
            seat.setSeatID(i);
            seats.add(seat);
        }
        // 60 to 100
        for (int i = 60; i<=100; i++){
            Seat seat = new Seat();
            seat.setSeatType(SeatType.GOLD);
            seat.setSeatID(i);
            seats.add(seat);
        }
        return seats;
    }

    private List<Screen> createScreen(){
        List<Screen> screens = new ArrayList<>();
        Screen screen1 = new Screen();
        screen1.setScreenID(1);
        screen1.setSeats(createSeats());
        screens.add(screen1);

        return screens;

    }

    private Show createShow(int showID, Screen screen, Movie movie, int showStartTime){
        Show show = new Show();
        show.setShowID(showID);
        show.setToTimeStamp(showStartTime);
        show.setMovie(movie);
        show.setScreen(screen);
        return show;
    }

    private void createTheatre() {

        Movie movie1 = movieController.getMovieByName("movie1");
        Movie movie2 = movieController.getMovieByName("movie2");
         

        Location lucknow = createLocation(123, "Jankipuram", "Lucknow", "UP", "INDIA");
        Location bangalore = createLocation(453, "Whitefield", "Bangalore", "Karnatak", "INDIA");
     
        Theatre inox = new Theatre();
        inox.setScreens(createScreen());
        inox.setTheatreID(0);
        inox.setLocation(bangalore);

        Show inoxShowMorning = createShow(1, inox.getScreens().get(0), movie1, 10);
        Show inoxShowEvening = createShow(2, inox.getScreens().get(0), movie2, 18);

    
        List<Show> inoxShows = new ArrayList<>();
        inoxShows.add(inoxShowEvening);
        inoxShows.add(inoxShowMorning);
        inox.setShows(inoxShows);

        theaterController.addTheatre(bangalore, inox);
    }


    private void initialize() {

        //create movies
        createMovies();

        //create theater with screens, seats and shows
        createTheatre();
    }

    public static void main(String args[]){
        BookMyShow bookMyShow = new BookMyShow();
        
        bookMyShow.initialize();
        Location bangalore = bookMyShow.createLocation(453, "Whitefield", "Bangalore", "Karnatak", "INDIA");
        bookMyShow.createBooking(bangalore, "movie1");

    }

}

File: LLD(Python)/bookMyShow/bookMyShow.py
from Controller.MovieController import MovieController
from Controller.TheatreController import TheatreController
from Model.Movie import Movie
from Model.Location import Location
from Model.Theatre import Theatre
from Model.Hall import Hall
from Model.Seat import Seat
from Model.Show import Show
from Model.Booking import Booking

from Enums.SeatType import SeatType

class BookMyShow:
    def __init__(self):
        # call controllers
        self.movieController = MovieController()
        self.theatreController = TheatreController()

    def createSeats(self)->list:
        seats = []
        for i in range(50):
            if (i<20):
                seat = Seat()
                seat.setSeatID(i)
                seat.setSeatType(SeatType.SILVER)
                seat.setRowno(i)
            elif (i>=20 and i<=40):
                seat = Seat()
                seat.setSeatID(i)
                seat.setSeatType(SeatType.GOLD)
                seat.setRowno(i)     
            else:    
                seat = Seat()
                seat.setSeatID(i+100)
                seat.setSeatType(SeatType.PLATINUM)
                seat.setRowno(i) 
            seats.append(seat)
        return seats

    def createShow(self, hall, movie, startTime, endTime):
        show = Show()
        show.setMovie(movie)
        show.setHall(hall)
        show.setStartTime(startTime)
        show.setEndTime(endTime)   
        return show   

    def createTheatre(self):
        avengers = self.movieController.getMovieByName("Avengers")
        rchock = Location("Rajiv Chock")
        delhi = Location("Delhi")
        inox = Theatre()

        inox.setAddress(rchock)
        inox.setCity(delhi)
        inoxHall1 = Hall("inoxHall1")
        # set seats for inoxHall1
        inoxHall1.setSeats(self.createSeats())
        inox.setHalls(inoxHall1)

        # inox.setMovieList(avengers)
        
        # add show
        inoxMorningShow = self.createShow(inox.getHalls()[0], avengers, 10, 20)
        inox.setShows(inoxMorningShow)
        self.theatreController.setTheatreByCity(delhi, inox)
        self.theatreController.setTheatreList(inox)

    def createMovies(self):
        avengers = Movie()
        avengers.setMovieName("Avengers")
        avengers.setMovieDuration(150)

        batman = Movie()
        batman.setMovieName("Batman")
        batman.setMovieDuration(180)

        delhi = Location("Delhi")
        mumbai = Location("Mumbai")
        self.movieController.addMovie(delhi, avengers)
        self.movieController.addMovie(mumbai, batman)

    def createBooking(self, city, movieName):
        # 1) Get movies by Location
        # 2) select the movies you want to setEndTime
        # 3) select the theatre you want to see the movie at
        # 4) select the show (slot at which you want to see the movie)
        # 5) see the available seats
        # 6) book the seat if you can book
        # 7) do the payment and get the receipt

        movieList = self.movieController.getMoviesByCity(city)
        desiredMovie = None
        for movie in movieList:
            if movie.getMovieName()==movieName :
                desiredMovie = movie
        
        desiredShow = None
        for theatre in self.theatreController.getTheatreList():
            for show in theatre.getShows():
                if show.getMovie().getMovieName()==desiredMovie.getMovieName():
                    desiredShow = show

        pickedSeat = 10
        if pickedSeat not in desiredShow.getBookedSeats():
            for seat in desiredShow.getHall().getSeats():
                if pickedSeat != seat.getSeatID():
                    pass
                else:
                    show.setBookedSeats(pickedSeat)
                    booking = Booking()
                    booking.setBookedSeats(pickedSeat)
                    booking.setShow(desiredShow)
                    booking.setBookedSeats(seat)
                    print("BOOKING IS SUCCESSFUL")
        else:
            print("SEAT IS ALREADY BOOKED")
                

    def initialize(self):
        self.createMovies()
        self.createTheatre()
        
if __name__ == '__main__':
    bookmyshow = BookMyShow()
    bookmyshow.initialize()
    delhi = Location("Delhi")
    bookmyshow.createBooking(delhi, "Avengers")

    # bookmyshow.createMovies()
    # bookmyshow.createTheatre()